### DISCUSSION TITLE
MicroPython can connect with supabase with urequest or mqtt?

### DISCUSSION THREAD

**Discussion by LautaroZanzot** [supabase/supabase#27214] [Questions] [answered]
I am working on my project with an esp8266 and when I try to make a POST with urequest it does not recognize the apikey, error 401, so can I log in through mqtt and make insters to the db?

**Comment by GaryAustin1** [comment]
You probably need to provide more info on what your call looks like an how you are providing the apikey header (and Authorization header if doing database operations).

**Reply by LautaroZanzot** [reply]
def function
`
json_data = ujson.dumps({
    "temp": str(tempC),
    "defrost": str(defrost),
    "arranque_comp": str(contrl),
    "bateria": str(bateria),
    "idvigia": str(IdVigia)
})

headers = {
    "Content-Type": "application/json",
    "Authorization": "Bearer " + API_KEY
}
response = request.post(API_URL + "/" + TableName, headers=headers, data=json_data)
if response.status_code == 401:
    print("Error 401: Unauthorized")
    print(API_URL + "/" + TableName, headers, json_data)
    print("Reason:", response.reason)
    print("Text:", response.text)
else:
    print("HTTP response code:", response.status_code)
    print(json_data)
    print(API_URL + "/" + TableName, headers, json_data)
`
    the result is
    Error 401: Unauthorized https://...supabase.com/signals {'Authorization': 'Bearer eyJh...3gc', 'Content-Type': 'application/json'} {"temp": "[]", "defrost": "0", "bateria": "0", "idvigia": "1", "arranque_comp": "0"} Reason: b'Unauthorized' Text: {"message":"No API key found in request","hint":"No apikey request header or url param was found."}

**Reply by GaryAustin1** [reply]
You need an apikey header also with the anon (or service_role) key (jwt).  No 'bearer' in it.

**Reply by LautaroZanzot** [reply]
is the same, like Authorization: API_SER its 401 and apikey: API_SER its 404

**Reply by GaryAustin1** [reply]
You need both headers.   And Authorization is 'Bearer jwt' like you showed in your first code.

**Reply by LautaroZanzot** [reply]
Sorry for the innocence, but it still gives an error, it has to be something like this, I attached the example
`    headers = {
        "Content-Type": "application/json",
        "Authorization": "Bearer " + API_SER,
        "api key": API_SER
    }`
 i tryed API_KEY variations too

**Reply by GaryAustin1** [reply]
`apikey`, not `api key`...

I don't know anything about your environment, but the error seems pretty clear.

**Reply by LautaroZanzot** [reply]
the space between was my fault writing here. my code is this. The device have to read the state of the GPIOs and send it to supabase:
`import network
import time
import urequests as request
import machine
import onewire
import ds18x20
import ujson


led = machine.Pin(16, machine.Pin.OUT)
oneWireBus = machine.Pin(2)
sensor = ds18x20.DS18X20(onewire.OneWire(oneWireBus))
pinDefrost = machine.Pin(4, machine.Pin.IN)
pinContrl = machine.Pin(5, machine.Pin.IN)
pinBateria = machine.Pin(13, machine.Pin.IN)


gc.collect()

SSID = 'SSID'
PWD = 'Passwd'
API_URL = "https://.supabase.co"
API_KEY = "eyJhb...3IMrV3gc"
API_SER = "eyJhb...PXg"
TableName = "signals"
httpsPort = 443
IdVigia = 1


def connectWifi(ssid, pwd):
    
    station = network.WLAN(network.STA_IF)
    station.active(True)
    station.connect(ssid, pwd)
    
    while not station.isconnected():
        print('conectando...')
        time.sleep(1)
        
    print(f'Conexion exitosa a {ssid}')



def send_data(tempC, defrost, contrl, bateria):
    json_data = ujson.dumps({
        "temp": str(tempC),
        "defrost": str(defrost),
        "arranque_comp": str(contrl),
        "bateria": str(bateria),
        "idvigia": str(IdVigia)
    })
    
    headers = {
        "Content-Type": "application/json",
        "Authorization": "Bearer " + API_SER,
        "apikey": API_SER
    }
    
    response = request.post(API_URL + "/" + TableName, headers=headers, data=json_data)
    if response.status_code == 401:
        print("Error 401: Unauthorized")
        print("Headers:", headers)
        print("Reason:", response.reason)
        print("Text:", response.text)
    else:
        print("HTTP response code:", response.status_code)
        print(json_data)
        print("URL:", API_URL + "/" + TableName)
        print("Headers:", headers)
        print("JSON Data:", json_data)


def main():

    connectWifi(SSID, PWD)
    while True:
        tempC = sensor.scan()
        defrost = pinDefrost.value()
        contrl = pinContrl.value()
        bateria = pinBateria.value()
            # Read the voltage levels
        voltage4 = pinDefrost.value()
        voltage5 = pinContrl.value()
        voltage13 = pinBateria.value()
        if voltage4 <= 1:
            defrost = 0
        else:
            defrost = 1
        if voltage5 <= 1:
            contrl = 0
        else:
            contrl = 1
        if voltage13 <= 1:
            bateria = 0
        else:
            bateria = 1
        print("tempC:", tempC)
        print("defrost:", defrost)
        print("arranque_comp:", contrl)
        print("bateria:", bateria)
        send_data(tempC, defrost, contrl, bateria)
        led.on()
        time.sleep(1)
        led.off()
        time.sleep(60)

main()
`

**Reply by GaryAustin1** [reply]
What is your current error with both?
Is API_SER service_role key from the dashboard? 
If it is anon key and your error changed then maybe RLS.

**Reply by LautaroZanzot** [reply]
RLS is disabled 
API_KEY = API KEY
API_SER = api service role
VIG_KEY = accestoken for the user in the device.
If I don't use the apikey tag, it says 401 apikey was not found, if I use apikey with VIG_KEY it says 401 check your api key or anon key in supabase, if I use API_KEY AND API_SER together it gives me a 404 error

**Reply by GaryAustin1** [reply, marked as answer]
apikey and Authorization header should be the same jwt. Either service_role together or anon key together.

Not sure what API_KEY = API KEY means.  You certainly would not use it with what ever vig_key is.

A 404 error means your URL is not found but did clear permission.  You should have said you got a different error.

I don't think you are formatting your URL correctly.

![image](https://github.com/supabase/supabase/assets/54564956/274b4a26-57b9-43c2-a224-e883da65063d)
https://supabase.com/docs/guides/api/quickstart?queryGroups=database-method&database-method=sql&queryGroups=language&language=jsz

In your dashboard there is an API tab and you can get all of your calls (BASH toggle for http requests) for each table.
https://supabase.com/dashboard/project/_/api

**Reply by LautaroZanzot** [reply]
Man, sorry for the strugle te problem was in the url, i love you
